### DISCUSSION TITLE
LOGIN does not work after upgrading to Airflow 2.7.3 "ERROR - Error returning OAuth user info: 'id_token'"

### DISCUSSION THREAD

**Discussion by piotrek9721** [apache/airflow#35939] [General]
### Apache Airflow version

2.7.3

### What happened

Hello everyone,
after the upgrade from Airflow 2.6.3 to 2.7.3 we can not login into the Airflow UI
In the webserver logs we encounter the error message 'ERROR - Error returning OAuth user info: 'id_token'' during the login process.
On the UI there is a info message with:
Invalid login. Please try again.

### Additional Information
- Airflow Version: 2.7.3
- Previous Airflow Version: 2.6.3
- OAuth
- Any relevant configuration changes made during the upgrade

### What you think should happen instead

Login should work without any errors, as it did in Airflow 2.6.3.

### How to reproduce

1. Upgrade Airflow to version 2.7.3.
2. Attempt to log in to the webserver.
3. Observe the error in the logs.

### Operating System

Linux

### Deployment

Official Apache Airflow Helm Chart

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template! If you are willing to raise PR to address this issue please do so, no need to wait for approval.

**Comment by potiuk** [comment]
This issue has no information on what kind of authentication you have, how you have it confgured etc. etc. - it' more of a troubleshooting and it is impossible to help you without any details .Logs from what's happening  would - for example - possibly help somoene who might attempt to help you in your problem.

**Comment by sungwy** [comment]
I'm running into a similar issue @piotrek9721 and @potiuk and I think it may be related to the AirflowSecurityManager class and custom user_info.

In our userinfo returned from the OAuth endpoint, we don't have the right attribute names, and have to fetch the values and create a new user_info dictionary that correctly maps the returned attributes to the expected FAB OAuth attributes (e.g. emailAddress to email)

It looks like using the same AirflowSecurityManager implementation is causing the webserver to not be able to find the expected attributes anymore:

`ERROR - OAUTH userinfo does not have username or email {}`

I'm observing above issue when upgrading from version 2.6.2 to 2.7.3

**Comment by potiuk** [comment]
I think it's the issue with FAB 4.3.9 -  https://github.com/dpgaspar/Flask-AppBuilder/pull/2163 it is going to be fixed in the upcoming Airflow 2.8.0 by upgrading to FAB 4.3.10 0 - It's been merged yesterday https://github.com/apache/airflow/pull/35991

**Comment by mvitale-kensu** [comment]
We have the same issue as @syun64: 

`ERROR - OAUTH userinfo does not have username or email {}`

But in our case we are upgrading from `2.7.2` to `2.8.1` which is using FAB 4.3.10

OAUTH provider configuration is unchanged and it's working fine if we go back to 2.7.2 - also no errors reported on the OAUTH provider logs. 
Unfortunately other than the error above there is nothing else in airflow logs, I tried to increase the log level to DEBUG but nothing.

**Comment by potiuk** [comment]
You need to increase FAB logging level https://github.com/apache/airflow/blob/main/airflow/config_templates/airflow_local_settings.py#L34C1-L35C1

**Comment by beatstream69** [comment]
After some investigation I rewrote my webserver_config.py and custom authorizer for Keycloak, hope this helps someone
<details><summary>webserver_config.py</summary> 

```python 
import os
from flask_appbuilder.security.manager import AUTH_OAUTH
from oauth_authorizer import KeycloakRoleAuthorizer


basedir = os.path.abspath(os.path.dirname(__file__))

WTF_CSRF_ENABLED = True
WTF_CSRF_TIME_LIMIT = None

AUTH_TYPE = AUTH_OAUTH
AUTH_ROLES_SYNC_AT_LOGIN = True  # Checks roles on every login
AUTH_USER_REGISTRATION = True  # allow users who are not already in the FAB DB to register
SECURITY_MANAGER_CLASS = KeycloakRoleAuthorizer
AUTH_ROLES_MAPPING = {
    "Admin": ["Admin"],
    "User": ["User"],
    "Viewer": ["Viewer"],
}
oauth_client_secret = os.getenv("OAUTH_CLIENT_SECRET")
# this allows mounting from k8s secret in helm config
oauth_client_secret_fn = os.getenv("OAUTH_CLIENT_SECRET_FILE", '/var/tmp/ocs')
if os.path.isfile(oauth_client_secret_fn):
    with open(oauth_client_secret_fn, 'r') as ocs: oauth_client_secret = ocs.read()
OAUTH_PROVIDERS = [
    {
        "name": "kc2fa",
        "icon": "fa-key",
        "token_key": "access_token",
        "remote_app": {
            "client_id": os.getenv("OAUTH_CLIENT_ID"),
            "client_secret": oauth_client_secret,
            "api_base_url": os.getenv("OAUTH_CLIENT_API_BASE_URL"),
            "access_token_url": os.getenv("OAUTH_ACCESS_TOKEN_URL"),
            "authorize_url": os.getenv("OAUTH_AUTHORIZE_URL"),
            "jwks_uri": os.getenv("OAUTH_JWKS_URL"),
            "request_token_url": None,
            "client_kwargs": {
                "scope": "email profile"
            },
        },
    },
]

# ----------------------------------------------------
# Theme CONFIG
# ----------------------------------------------------
# Flask App Builder comes up with a number of predefined themes
# that you can use for Apache Airflow.
# http://flask-appbuilder.readthedocs.io/en/latest/customizing.html#changing-themes
# Please make sure to remove "navbar_color" configuration from airflow.cfg
# in order to fully utilize the theme. (or use that property in conjunction with theme)
# APP_THEME = "bootstrap-theme.css"  # default bootstrap
# APP_THEME = "amelia.css"
# APP_THEME = "cerulean.css"
# APP_THEME = "cosmo.css"
# APP_THEME = "cyborg.css"
APP_THEME = "darkly.css"
# APP_THEME = "flatly.css"
# APP_THEME = "journal.css"
``` 
</details> 


<details><summary>oauth_authorizer.py</summary> 

```python 
from flask import redirect
from flask_appbuilder import expose
from flask_appbuilder.security.views import AuthOAuthView
from flask_login import logout_user
from airflow.auth.managers.fab.security_manager.override import FabAirflowSecurityManagerOverride
from typing import Any, List, Union
import logging
import os
import jwt


log = logging.getLogger(__name__)
log.setLevel(os.getenv('AIRFLOW__LOGGING__FAB_LOGGING_LEVEL', 'INFO'))

FAB_ADMIN_ROLE = 'Admin'
FAB_USER_ROLE = 'User'
FAB_VIEWER_ROLE = 'Viewer'
FAB_PUBLIC_ROLE = 'Public'  # The 'Public' role is given no permissions

KC_ADMIN_ROLE = 'AirflowAdmin'
KC_USER_ROLE = 'AirflowUser'
KC_VIEWER_ROLE = 'AirflowViewer'


def map_roles(team_list: list) -> list:
    team_role_map = {
        KC_ADMIN_ROLE: FAB_ADMIN_ROLE,
        KC_USER_ROLE: FAB_USER_ROLE,
        KC_VIEWER_ROLE: FAB_VIEWER_ROLE,
    }
    return list(set(team_role_map.get(team, FAB_PUBLIC_ROLE) for team in team_list))


class KeycloakAuthRemoteUserView(AuthOAuthView):

    @expose("/logout/")
    def logout(self):
        logout_user()
        return redirect(os.getenv("OAUTH_LOGOUT_REDIRECT_URL"))
# OAUTH_LOGOUT_REDIRECT_URL = "https://kc.example.com/realms/myrealm/protocol/openid-connect/logout?client_id=airflow&post_logout_redirect_uri=https%3A%2F%2Fairflow.example.com%2F"

class KeycloakRoleAuthorizer(FabAirflowSecurityManagerOverride):

    authoauthview  = KeycloakAuthRemoteUserView

    def get_oauth_user_info(self, provider: str, resp: Any) -> dict:
        access_token = resp.get('access_token')
        #log.info(f"Token: {access_token}")
        token_data = jwt.decode(access_token, options={"verify_signature": False})
        #log.info(f"Token data: {token_data}")
        roles = map_roles(token_data['realm_access']['roles'])
        return {'username': token_data.get('preferred_username'),
                'first_name': token_data.get('given_name'),
                'last_name': token_data.get('family_name'),
                'email': token_data.get('email'),
                'role_keys': roles}
```
</details>

**Reply by PIrojahPerbak** [reply]
Worked for. Just upgraded to airflow 2.10 from 2.6
